### ISSUE TITLE
LSP: big workspace causes "subcapture nesting too deep" error in Lpeg.P:match

### ISSUE THREAD

**Issue by fegyi001** [neovim/neovim#26520]
### Problem

Really strange error, and I just could not find anything online for solving that.
If I'm into a pretty large codebase (Angular monorepo) and try to save a file after editing, sometimes the following error occures and I'm unable to save:

```bash
Error detected while processing BufWritePre Autocommands for "*":
Error executing lua callback: ...vim-macos/share/nvim/runtime/lua/vim/lsp/_watchfiles.lua:85: subcapture nesting too deep
```

<img width="1515" alt="image" src="https://github.com/neovim/neovim/assets/4125222/a46f5c5a-4820-4b8a-a53b-4c62a8f787f7">

If I understand it correctly the autoformatter cannot do its stuff, but I'm not sure. I'm not a neovim ninja, therefore any help is highly appreciated.

### Steps to reproduce using "nvim -u minimal_init.lua"

```lua
--- CHANGE THESE
local pattern = 'the-filetype'
local cmd = {'name-of-language-server-executable'}
-- Add files/folders here that indicate the root of a project
local root_markers = {'.git', '.editorconfig'}
-- Change to table with settings if required
local settings = vim.empty_dict()

vim.api.nvim_create_autocmd('FileType', {
  pattern = pattern,
  callback = function(args)
    local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
    local root_dir = match and vim.fn.fnamemodify(match, ':p:h') or nil
    vim.lsp.start({
      name = 'bugged-ls',
      cmd = cmd,
      root_dir = root_dir,
      settings = settings
    })
  end
})
```

### Expected behavior

the file should be saved and formatted as usual

### Neovim version (nvim -v)

v0.10.0-dev-1801+g2ebd328a7

### Language server name/version

typescript

### Operating system/version

MacOS 14.0

**Comment by glepnir** [comment]
relate https://github.com/neovim/neovim/pull/26108

**Comment by mfussenegger** [comment]
This sounds like an issue with the globbing function/lpeg pattern.

Could you share reproduction steps that one can follow to reproduce the issue?

**Comment by ChilloManiac** [comment]
Im seeing this error too - Also running on MacOs

**Comment by fegyi001** [comment]
there is not much to reproducing it. I just try to save a file in the codebase. It occurs when the file is deep in the file tree.
With the workaround now the error disappeared, so thanks for that!

**Comment by j1ah0ng** [comment]
I am also encountering this error on file open, albeit on a _very_ large (100s of GBs) monorepo with `rust-analyzer`. Tailing `:LspLog` gives

```
[ERROR][2024-02-09 23:06:30] .../lua/vim/lsp.lua:852    "LSP[rust_analyzer]"    "on_error"  {  code = "SERVER_REQUEST_HANDLER_ERROR",  err = "/usr/share/nvim/runtime/lua/vim/glob.lua:79: subcapture nesting too deep"}
```

which points to a 
```
local lpeg_pattern = Lpeg.P:match(pattern)
```
call.

**Comment by j1ah0ng** [comment]
I spent some more time poking around and found [this reference to the error message](https://github.com/edubart/nelua-lang/blob/master/src/lpeglabel/lplcap.c#L457) in `lplcap.c` within the LPeg source: 
```c
static int pushcapture (CapState *cs) {
  lua_State *L = cs->L;
  int res;
  luaL_checkstack(L, 4, "too many captures");
  if (cs->reclevel++ > MAXRECLEVEL)
    return luaL_error(L, "subcapture nesting too deep");

```
there appears to be a hard-coded `#define MAXRECLEVEL‎ 200` which appears to correspond to the Lua interpreter's max recursion level ([for reference, a -DMAXRECLEVEL compiler flag](https://github.com/edubart/nelua-lang/blob/05a2633a18dfdde7389394b9289da582c10e79bc/Makefile#L32)).

A cursory search of the neovim repo did not give me a satisfying answer as to what the max depth of the bundled interpreter is: I'm also not sure how LPeg is bundled into neovim, so I'm not certain what the correct way of bumping this number would be.

**Comment by lewis6991** [comment]
Bumping the number isn't going to be an acceptable solution. We need to understand why it is hitting this limit.

**Comment by j1ah0ng** [comment]
I may also be totally off base: found [this](https://neovim.io/doc/user/lua.html#vim.lpeg.setmaxstack()) reference to a `max stack` used by LPeg in particular:
```
Returns a pattern that matches any single character that appears in the given string (the S stands for Set). As an example, the pattern lpeg.S("+-*&zwj;/") matches any arithmetic operator. Note that, if s is a character (that is, a string of length 1), then lpeg.P(s) is equivalent to lpeg.S(s) which is equivalent to lpeg.R(s..s). Note also that both lpeg.S("") and lpeg.R() are patterns that always fail.
```

**Comment by j1ah0ng** [comment]
> Bumping the number isn't going to be an acceptable solution. We need to understand why it is hitting this limit.

My suspicion is that the deeply nested directory structure present in my repo is causing some OOB check to fail. 

I also want to point out that my use case is particular nasty given that my working directory is ~5 levels deep within a source tree, and that my `compile_commands` points to codegen source files which live 10s of levels deep within Bazel build cache indirection.

**Comment by j1ah0ng** [comment]
Making the change to `MAXRECDEPTH` appears to have gotten rid of the issue: [diff](https://github.com/j1ah0ng/neovim/commit/8f5b977ea406af46375b4a6cc63b8b285506c58e). 

Placing a `print(debug.traceback())` before the offending line in `glob.lua` seems to substantiate something is going wrong with watching deeply-nested source trees:
```
stack traceback:
^I/usr/share/nvim/runtime/lua/vim/glob.lua:79: in function 'to_lpeg'
Press ENTER or type command to continue
^I/usr/share/nvim/runtime/lua/vim/lsp/_watchfiles.lua:65: in function 'register'
Press ENTER or type command to continue
^I/usr/share/nvim/runtime/lua/vim/lsp/handlers.lua:123: in function </usr/share/nvim/runtime/lua/vim/lsp/handlers.lua:110>
Press ENTER or type command to continue
^I[C]: in function 'pcall'
```

**Comment by j1ah0ng** [comment]
Actually instrumenting the pattern also appears to substantiate this:
```
/home/jiahong.long/.cache/bazel/_bazel_jiahong.long/b83d1475e1d59eed9b220ab04e73408d/execroot/cruise_ws/bazel-out/k8-fastbuild/bin/tools/rust/examples/bag_reading/bag_reading_build_script.out_dir/**/*.rs
```

**Comment by BitInByte** [comment]
I’ve also experienced this issue yesterday for the first time. After creating a new directory with some files in it (using oil.nvim if it helps), that files reproduced this error when I tried to save (had formatter and eslint running on save as well on save). Probably I was really deep in sub directories but was really strange. The only way to make the formatters work again was to delete that directory and move the files inside to the parent directory.

haven’t experienced this before and started some days after updating to version 0.10 but don’t know if it’s derived from this update or if it was because it was really deep into multiple sub directories or even something else

**Comment by BitInByte** [comment]
Just wondering why was my comment marked as duplicated when I have some new insights of steps that caused the issue but yeah, hope team will be able to find a fix

**Comment by justinmk** [comment]
Your comment was marked as duplicate because it does not add useful info. That isn't a prompt to add even more comments.

**Comment by BitInByte** [comment]
What I am trying to understand is, is this ever gonna be considered to be fixed or not? As far as I can see, it still has the label that needs:repro and when a user comments with new insights (don’t know if you read my comment at least but I reference new stuff in there) you just mark as duplicate and even, I can see that this issue has been referenced in some other commits, my question here is, how can that commits reference this issue if, as label indicates, can’t be reproduced? What is the metric here then? Is is reproducible or not in the end since it has been referenced in some other fixes?

don’t get me wrong, just want to understand if this will ever be considered or not because I am not the only one reporting and this is kind of a big issue for some people.

thanks and hope team makes neovim even bigger <3

**Comment by lewis6991** [comment]
> What I am trying to understand is, is this ever gonna be considered to be fixed or not?

The issue is still open.

> As far as I can see, it still has the label that needs:repro and when a user comments with new insights (don’t know if you read my comment at least but I reference new stuff in there) you just mark as duplicate and even, I can see that this issue has been referenced in some other commits, my question here is, how can that commits reference this issue if, as label indicates, can’t be reproduced? What is the metric here then? Is is reproducible or not in the end since it has been referenced in some other fixes?

Reproduction steps have yet to be provided. Your comment did not provide any, nor any useful information.

If a dev can't reproduce the issue, they cannot fix it. I don't know what is difficult to understand about that.

For anyone reading please refrain from any comments like:
- "I'm seeing this issue too"
- "This didn't happen on version 0.x"
- "When running this plugin (with this plugin)^n, the issue occurs".

Please **DO** comment if you can provide precise steps in the form of commands with a minimal init.lua that produces the issue., **just as the issue template instructs**.

**Comment by nyz93** [comment]
I've created a reproduction:
```lua
local cap = {
  textDocumentSync = {
    openClose = true,
  },
  textDocument = {
    formatting = { dynamicRegistration = true },
  },
}

local start = function(disp)
  local msg = 1
  local function request(mth, p, cb)
    msg = msg + 1
    if mth == "initialize" then
      cb(nil, { capabilities = cap })
    end
    return true, msg
  end

  local function notify(mth, p)
    if mth == "textDocument/didOpen" then
      disp.server_request("client/registerCapability", {
        registrations = {
          {
            id = "reg1",
            method = "textDocument/formatting",
            registerOptions = {
              documentSelector = {
                {
                  pattern = vim.fn["repeat"]("a", 201),
                },
              },
            },
          },
        },
      })
    end
  end

  return {
    request = request,
    notify = notify,
    is_closing = function()
      return false
    end,
    terminate = function() end,
  }
end

vim.api.nvim_create_autocmd("FileType", {
  pattern = "javascript",
  callback = function(args)
    vim.lsp.start({
      name = "bugged-ls",
      cmd = start,
    })
  end,
})

```
To trigger the bug just open a `.js` file and call `vim.lsp.buf.format()`.

The problem is that `to_lpeg` cannot handle long patterns, every character causes an additional recursion level, and eslint for example uses `client/registerCapability` to add itself as a formatter constrained to a "pattern" of the whole file path which can easily exceed 200 characters in bigger projects.

I've tried to debug `to_lpeg`, so far without success. The glob grammar can be reduced to this and it still fails on bigger patterns:
```lua
local p = P({
  "Pattern",
  Pattern = V("Elem") ^ -1 * V("End"),
  Elem = Cf(V("Literal") * (V("Elem") + V("End")), mul),
  Literal = P(1) / P,
  End = P(-1) * Cc(P(-1)),
})
```
The issue is most likely the definition of `Elem` and `Literal`, repeating `P(1) / P` is what causes the recursion to grow. The fix would be to rewrite the grammar to match all the consecutive literal characters and then transform into a pattern instead of doing it character by character, but I don't know how to do that.

**Comment by nyz93** [comment]
Wrapping up the captures containing `Literal` in a `Cmt` calls solves the issue but I don't know if there are downsides to using it.

**Comment by BitInByte** [comment]
Hey guys, I am facing this exact same issue still on Neovim 0.10.1. To trigger this error, everytime that I access a file which has a big path, mainly on angular projects, if I try to format (happen either with null-ls and conform), the format just throws this same error and from this point on, it will always throw this error whenever I try to format any other file, independently of the size of the path) and if I just exit and enter neovim, I can format unless I visit that big path again. Any help is much appreciated since I really need to use a formatter. Thanks in advance for any incoming help

**Comment by justinmk** [comment]
This was fixed in 0.11
